architecture/replication-layer.md

```diff
@@ -0,0 +1,77 @@
+
+Up to F failures can be tolerated, where the total number of replicas N = 2F + 1 (e.g. with 3x replication, one failure can be tolerated; with 5x replication, two failures, and so on).
+
+However, the number of replicas is ultimately controlled with table-level granularity in your cluster's zone config file.
+
+CockroachDB ensures consistency on these replicas––despite being distributed––by using the Raft consensus algorithm.
+
+## Components
+
+### Raft
+
+Raft works by organizing all nodes that contain a replica of a range into a group––unsurprisingly called a Raft Group.
+
+Raft elects a relatively long-lived leader which must be involved to propose commands. It heartbeats followers periodically and keeps their logs replicated. In the absence of heartbeats, followers become candidates after randomized election timeouts and proceed to hold new leader elections. Cockroach weights random timeouts such that the replicas with shorter round trip times to peers are more likely to hold elections first (not implemented yet). Only the Raft leader may propose commands; followers will simply relay commands to the last known leader.
+
+Once a node receives `BatchRequests` for a range it contains, it converts those KV operations into Raft commands. Those commands are proposed to the Raft group leader––which is what makes it ideal for the Lease Holder and the Raft leader to be one in the same; it reduces network roundtrips if the Lease Holder receiving the requests can simply propose the Raft commands to itself, rather than communicating them to another node.
+
+Proposed Raft commands are proposed to have them serialized through quorum voting into a coherent distributed log.
+
+#### Logs
+
+The log voted on by the Raft group contains an ordered set of commands that a quorum of replicas acknowledged and can be replayed to bring the node from a past state to its current state.
+
+This log also lets nodes that temporarily went offline to be "caught up" to the current state without needing to receive a copy of the existing data in the form of a snapshot.
+
+### Snapshots
+
+Each replica can be "snapshotted", which copies all of its data as of a specific MVCC timestamp. This snapshot can be sent to other nodes during a rebalance event to expedite replication.
+
+After receiving a snapshot, the node gets it up to date by playing all of the actions from the Raft group's committed log that have occurred since the snapshot was taken.
+
+### Leases
+
+Leases are held by a single member of the Raft group (usually the same node as the Raft group leader), and the leaseholder is responsible for receiving `BatchRequests` for all reads and writes related to keys in the range. This has the benefit of improving the speed of reads drastically (which don't need consensus, since a quorum of nodes already agreed with the leaseholder as to the status of the node after the write ).
```

Comment:
I think the definition is in there (they receive all reads and writes related to the keys in the range). But I think the def and value need to be explained a bit more directly up-front.